lib/strCmp.man
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "strCmp 3"
.TH strCmp 3 "2013-09-07" "perl v5.18.1" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
.PP
\fIFind literally a string into an other\fR
.IX Subsection "Find literally a string into an other"
.PP
.Vb 1
\&        fndLiteral(string: string_to_find, string: string_were_find)
\&
\&        ex:
\&        my $found = fndLiteral("llo", "Hello world");
\&
\&        NOTE: 
\&        Return 1 if string is found into the other string
.Ve
.PP
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
.PP
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
.PP
\fIFind literally a string into an other but non apha-numeric charset are removed\fR
.IX Subsection "Find literally a string into an other but non apha-numeric charset are removed"
.PP
.Vb 1
\&        fndCollapse(string: string_to_find, string: string_were_find)
\&
\&        ex:
\&        my $found = fndCollapse("llo wor", "Hello \-\- world");
\&
\&        NOTE: 
\&        Return 1 if string is found into the other string
.Ve
.PP
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
.PP
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
.PP
\fIDamerau-Levenshtein distance computation\fR
.IX Subsection "Damerau-Levenshtein distance computation"
.PP
.Vb 1
\&        damerauLevenshteinLenght(string: string_to_find, string: string_were_find)
\&
\&        ex:
\&        my $dLenght = damerauLevenshteinLenght("llo wor", "Hello world");
\&
\&        NOTE: 
\&        Return (int) distance between two strings
.Ve
.PP
\&\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-


lib/strCmp.pm
#!/usr/bin/perl
#strCmp.pm
#
#  "Copyright 2013 Mauro Ghedin"
#
#  This program is free software; you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published by
#  the Free Software Foundation; either version 3 of the License, or
#  or any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details.
#
#  You should have received a copy of the GNU General Public License
#  along with this program; if not, write to the Free Software
#  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
#  MA 02110-1301, USA.
#
#       @author         : Mauro Ghedin
#       @contact        : domyno88 at gmail dot com
#       @version        : 0.1


use lib::base::base;

=pod

---------------------------------------------------------------------------------------------------

=head3 Find literally a string into an other

	fndLiteral(string: string_to_find, string: string_were_find)

	ex:
	my $found = fndLiteral("llo", "Hello world");

	NOTE: 
	Return 1 if string is found into the other string

---------------------------------------------------------------------------------------------------

=cut

sub fndLiteral ($$) {
	$str1 = shift;	#	string to find
	$str2 = shift;	#	string were find
	
	return ( $str2 =~ /$str1/ig ? 1 : 0);		#	return true if str1 is in str2
}

=pod

---------------------------------------------------------------------------------------------------

=head3 Find literally a string into an other but non apha-numeric charset are removed

	fndCollapse(string: string_to_find, string: string_were_find)

	ex:
	my $found = fndCollapse("llo wor", "Hello -- world");

	NOTE: 
	Return 1 if string is found into the other string

---------------------------------------------------------------------------------------------------

=cut

sub fndCollapse ($$) {
	$str1 = shift;	#	string to find
	$str2 = shift;	#	string were find
	
	$str1 =~ s/[^\w\d]//g;
	$str2 =~ s/[^\w\d]//g;
	
	return ( $str2 =~ /$str1/ig ? 1 : 0);		#	return true if str1 is in str2
}

=pod

---------------------------------------------------------------------------------------------------

=head3 Damerau-Levenshtein distance computation

	damerauLevenshteinLenght(string: string_to_find, string: string_were_find)

	ex:
	my $dLenght = damerauLevenshteinLenght("llo wor", "Hello world");

	NOTE: 
	Return (int) distance between two strings

---------------------------------------------------------------------------------------------------

=cut

sub damerauLevenshteinLenght ($$) {
	my @arStr1 = strToArray(lc shift);
	my @arStr2 = strToArray(lc shift);
	
	
	my @DA;
	my @d;																							#	distance matrix
	
	my $cost = 0;
	my $strSum = scalar @arStr1 + scalar @arStr2;	
	
	$d[0][0] = $strSum;
	
	for my $indx (0..scalar @arStr1 + 1) {															#	Levenshtein matrix init ( x axe)
		$d[$indx + 1][1] =  $indx;
		$d[$indx + 1][0] =  $strSum;
	}
	
	for my $indx (0..scalar @arStr2 + 1) {															#	Levenshtein matrix init ( y axe)
		$d[1][$indx + 1] =  $indx;
		$d[0][$indx + 1] =  $strSum;
	}
	
	for my $index (0..94) { $DA[$index] = 0 };														# 94 : ascii printable char
	
	for my $i (1..scalar @arStr1 + 1) {
		my $DB = 0;
		print "\n";
		for my $j (1..scalar @arStr2 + 1) {
			
			my $i1 = $DA[$arStr2[$j - 1]];
			my $j1 = $DB;
			
			$cost = 1;
			
			$cost = 0 if $arStr1[$i - 1] eq $arStr2[$j - 1];
			
			$DB = $j if $cost == 0;
			
			$d[$i+1][$j+1] = min (	$d[$i][$j] + $cost, 
									$d[$i + 1][$j] + 1, 
									$d[$i][$j + 1] + 1, 
									$d[$i1][$j1] + ($i - $i1 - 1 ) + 1 + ($j - $j1 - 1 )
								);
								
			print $d[$i][$j]."   ";
		}
		$DA[$arStr1[$i - 1]] = $i;
	}
	return $d[scalar @d - 1][scalar @{ $d[0] } - 1];			
}	

1;
